make/jdk/src/classes/build/tools/taglet/PreviewNote.java

```diff
@@ -0,0 +1,127 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package build.tools.taglet;
+
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Set;
+
+
+import javax.lang.model.element.Element;
+import javax.tools.Diagnostic;
+
+
+import com.sun.source.doctree.DocTree;
+import com.sun.source.doctree.UnknownInlineTagTree;
+import jdk.javadoc.doclet.Doclet;
+import jdk.javadoc.doclet.DocletEnvironment;
+import jdk.javadoc.doclet.Reporter;
+import jdk.javadoc.doclet.StandardDoclet;
+import jdk.javadoc.doclet.Taglet;
+
+import static com.sun.source.doctree.DocTree.Kind.UNKNOWN_INLINE_TAG;
+
+/**
+ * An inline tag to insert a note formatted as preview note.
+ * The tag can be used as follows:
+ *
+ * <pre>
+ * {&commat;previewNote jep-number [Preview note heading]}
+ * Preview note content
+ * {&commat;previewNote}
+ * </pre>
+ *
+ */
+public class PreviewNote implements Taglet {
+
+    static final String TAG_NAME = "previewNote";
+    Reporter reporter = null;
+
+    @Override
+    public void init(DocletEnvironment env, Doclet doclet) {
+        if (doclet instanceof StandardDoclet stdoclet) {
+            reporter = stdoclet.getReporter();
+        }
+    }
+
+    /**
+     * Returns the set of locations in which the tag may be used.
+     */
+    @Override
+    public Set<Location> getAllowedLocations() {
+        return EnumSet.allOf(Taglet.Location.class);
+    }
+
+    @Override
+    public boolean isInlineTag() {
+        return true;
+    }
+
+    @Override
+    public String getName() {
+        return TAG_NAME;
+    }
+
+    @Override
+    public String toString(List<? extends DocTree> tags, Element elem) {
+
+        for (DocTree tag : tags) {
+            if (tag.getKind() == UNKNOWN_INLINE_TAG) {
+                UnknownInlineTagTree inlineTag = (UnknownInlineTagTree) tag;
+                String[] content = inlineTag.getContent().toString().trim().split("\\s+", 2);
+                if (!content[0].isBlank()) {
+                    StringBuilder sb = new StringBuilder("""
+                       <div class="preview-block" style="margin-top:10px; display:block; max-width:max-content;">
+                       """);
+                    if (content.length == 2) {
+                        sb.append("""
+                                <div class="preview-label">
+                                """)
+                          .append(content[1])
+                          .append("""
+                                </div>
+                                """);
+                    }
+                    sb.append("""
+                            <div class="preview-comment">
+                            """);
+                    return sb.toString();
+                } else {
```

Comment:
What does this else block do? In what case can the JEP number be blank?

Reply:
The else block generates the closing tag for the note. As you see in the PR description, the tag syntax resembles HTML open/close tags. This is a compromise to be able to locate the note within the description (which isn't possible with block tags), and being able to use rich content in the note without being required to writing a dedicated tag parser and rendering code. (I also happen to find this syntax easier to read than a custom inline tag, which would require additional rules to separate meta data from note content.)

Reply:
Such usage should be reflected in `testPreviewNoteTag`, which currently does not use it as open/close tags.

Reply:
The problem is the Taglet class is part of the build system, and I don't think we have a way to test these build classes (I coulndn't find any tests for the other build Taglets). I could replicate the Taglet class in the test, but that doesn't seem like a good solution either.

Reply:
Can we have a usage example of this tag somewhere then?

Reply:
There is a usage example in the Taglet class comment.